pages/api/companies.ts

```diff
@@ -1,5 +1,134 @@
+  [MessageType.COMPANY_CREATED_SUCCESSFULLY]: {
+    type: StatusMessageType.SUCCESS,
+    message: 'Company created successfully.',
+  },
+  [MessageType.EMPTY_NAME]: { type: StatusMessageType.ERROR, message: 'Company name is empty.' },
+  [MessageType.INVALID_COMPANY_URL]: { type: StatusMessageType.ERROR, message: 'Company url is invalid.' },
+  [MessageType.MISSING_NAME]: { type: StatusMessageType.ERROR, message: 'Company name is missing.' },
+  [MessageType.MISSING_COMPANY_URL]: { type: StatusMessageType.ERROR, message: 'Company url is missing.' },
+});
+
+function handler(req: NextApiRequest, res: NextApiResponse) {
+  const method = req.method;
+
+  switch (method) {
+    case HttpMethod.GET:
+      handleGet(req, res);
+      break;
+    case HttpMethod.POST:
+      handlePost(req, res);
+      break;
+    default:
+      rejectHttpMethod(res, method);
+  }
+}
+
+async function handleGet(req: NextApiRequest, res: NextApiResponse<ApiResponse<CompanyListData[]>>) {
+  const { name } = req.query;
+  const trimmedNameKeyword = name && name.toString().trim();
+
+  if (!trimmedNameKeyword) {
+    res.status(HttpStatus.OK).json(createJsonResponse([]));
+    return;
+  }
+
+  const companies: CompanyListData[] = await prisma.company.findMany({
+    where: {
+      isVerified: true,
+      name: {
+        contains: trimmedNameKeyword,
+        mode: 'insensitive',
+      },
+    },
+    orderBy: {
+      name: 'asc',
+    },
+    select: { id: true, name: true, companyUrl: true },
+  });
+
+  res.status(HttpStatus.OK).json(createJsonResponse(companies));
+}
+
+async function handlePost(req: NextApiRequest, res: NextApiResponse<ApiResponse<CompanyData>>) {
+  const errorMessageType = validateRequest(req);
+  if (errorMessageType !== null) {
+    res.status(HttpStatus.BAD_REQUEST).json(createJsonResponse({}, messages[errorMessageType]));
+    return;
+  }
+
+  const companyPostData: CompanyPostData = {
+    name: capitalizeEveryWord(req.body.name),
+    companyUrl: removeProtocolAndWwwIfPresent(req.body.companyUrl),
+  };
+
+  const duplicateCompany = await prisma.company.findFirst({
+    where: {
+      name: companyPostData.name,
+      companyUrl: companyPostData.companyUrl,
+    },
+    select: { id: true, name: true, companyUrl: true },
+  });
+
+  if (duplicateCompany) {
+    res.status(HttpStatus.OK).json(createJsonResponse(duplicateCompany, messages[MessageType.COMPANY_ALREADY_EXISTS]));
+    return;
+  }
+
+  const newCompany = await prisma.company.create({
+    data: {
+      name: companyPostData.name,
+      companyUrl: companyPostData.companyUrl,
+    },
+    select: { id: true, name: true, companyUrl: true },
+  });
+
+  res
+    .status(HttpStatus.CREATED)
+    .json(createJsonResponse(newCompany, messages[MessageType.COMPANY_CREATED_SUCCESSFULLY]));
+}
+
+function validateRequest(req: NextApiRequest): Nullable<MessageType> {
+  if (!req.body.name) {
```

Comment:
Empty `req.body.name` will drop into this if statement. We need to do `req.body.name === undefined`